### ISSUE TITLE
Most popular content is adding the full path twice for subdirectory installs (when a path is already defined within Analytics)

### ISSUE THREAD

**Issue by jamesozzie** [google/site-kit-wp#2821]
## Bug Description

If a user user sets up a WordPress install on a subdirectory (_mysite.com/blog_) and then connects to Analytics using a property that includes the full subdirectory in their Analytics "Default URL" field this results in invalid URLs specific links in the "Most popular content" widget (with the path referenced twice)

**_Example default URL including the path at Analytics level_**

![image](https://user-images.githubusercontent.com/41326532/108240856-27d5c280-7143-11eb-812f-f6f6cef2423b.png)

**_The "Most popular content" graph_**

![image](https://user-images.githubusercontent.com/41326532/108240380-b4cc4c00-7142-11eb-8e7d-3e75b0c80df4.png)

**_Final URL with path listed twice_**

![image](https://user-images.githubusercontent.com/41326532/108239411-c82ae780-7141-11eb-9f68-376a1d09e924.png)

## Steps to reproduce

1. Setup WordPress on a subdirectory 
2. Setup an Analytics property and include the subdomain in the "Default URL" field (A universal type property)
3. Connect the Analytics module within Site Kit (connect the same account and property)
4. Visit the "Most popular content" widget and click on any of the titles for more insights
5. Issue occurs, with the path listed twice

## Additional Context

 - SK version 1.16.0
 - This came up in a [recent support topic](https://wordpress.org/support/topic/post-titles-in-google-sitekit-poiting-to-wrong-url-on-my-website/)
 - One workaround is to remove the path from the Analytics property settings (at analytics.google.com)

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* When combining URL paths from an Analytics API report with the current site URL for a site that includes a path segment in its site URL, this should be done without duplicating the path segment.
    * Analytics URL paths always include the full path, so basically when concatenating, any path segment should be stripped from the current site URL, to ensure it's not included twice.
    * Potentially it's worth checking that the Analytics URL path actually starts with the correct path before "assuming" so.
* This needs to be fixed throughout various Analytics data tables in the plugin:
    * the WordPress dashboard widget
    * the Popular Pages widget in the Site Kit dashboard and the Analytics module page
    * the Top Earning Pages widget in the Site Kit dashboard and the AdSense module page (both require AdSense to be also connected and linked to Analytics)

## Implementation Brief

Create a new util function, it will be similar to the `getURLPath(path)` utility in the same folder, with the key difference that it returns the absolute url for a path including the siteURL.
`assets/js/util/getFullURL.js`

The function will take two params, `siteURL`, and `path`

The implementation will be the following:
```js
function getFullURL(siteURL, path) {
  return new URL(path, siteURL).href;
}
```

Update our code to use this function in the following two locations:
Legacy Data Table
 https://github.com/google/site-kit-wp/blob/31d77ec6a4370be08ec67177156479535237f8fa/assets/js/components/data-table.js#L53

DetailsPermalinks component (new approach)
https://github.com/google/site-kit-wp/blob/31d77ec6a4370be08ec67177156479535237f8fa/assets/js/components/DetailsPermaLinks.js#L39

### Test Coverage

* Create a corresponding test file in the util folder.

### Visual Regression Changes

* N/A (urls don't show).
 
## QA Brief

* Update your site to have a subfolder in the url `/wp-admin/options-general.php` e.g. `http://example.com/foo`, update the GA property to correspond to the same (see https://support.google.com/analytics/answer/3467852?hl=en ).
* Enable widgets
* Click through the links on `/wp-admin/admin.php?page=googlesitekit-dashboard` under the `Most popular content` section.
* Repeat the same for the widget version.
* Ensure you get actual data and not the error message as shown in the bug description `It looks like the URL is not part of this site, therefore there is no data available to display`.
* Ensure that the url in the analytics doesn't have the folder twice (not http://example.com/foo/foo/slug)

## Changelog entry

* Resolved analytics URL linking issue in subfolder based installs.

**Comment by jamesozzie** [comment]
@abdullah1908 Unsure whether this is something that can be addressed on the plugin side, or should users define the parent URL on their Analytics configuration.

**Comment by abdullah1908** [comment]
@jamesozzie I can reproduce this issue on my end as well.
![Screenshot (24)](https://user-images.githubusercontent.com/14175927/108266117-86bf2a00-718b-11eb-8aa4-debede7cf4e2.png)

**Comment by ivankruchkoff** [comment]
This is an issue in the legacy dashboard, but I don't think it is in the new widget approach:

In legacy, we have a siteURL (e.g. example.com/foo ), and the url returned from the report is relative to domain root, so it will start with /foo
https://github.com/google/site-kit-wp/blob/31d77ec6a4370be08ec67177156479535237f8fa/assets/js/components/data-table.js#L53

https://github.com/google/site-kit-wp/blob/31d77ec6a4370be08ec67177156479535237f8fa/assets/js/modules/analytics/components/dashboard/LegacyAnalyticsDashboardWidgetPopularPagesTable.js#L83

As returned URLs are relative to the domain root, we get the double folder portion in the url.

One potential fix is to strip the folder here: https://github.com/google/site-kit-wp/blob/31d77ec6a4370be08ec67177156479535237f8fa/assets/js/modules/analytics/components/dashboard/LegacyAnalyticsDashboardWidgetPopularPagesTable.js#L83

We can do this by checking whether the url starts with the folder portion of global._googlesitekitLegacyData.admin.siteURL ( https://developer.mozilla.org/en-US/docs/Web/API/URL ), and stripping it so a url like `/foo/bar-123` with a siteURL of `http://example.com/foo` will become `/bar-123`.

Something like this should work:
https://github.com/google/site-kit-wp/blob/31d77ec6a4370be08ec67177156479535237f8fa/assets/js/modules/analytics/components/dashboard/LegacyAnalyticsDashboardWidgetPopularPagesTable.js#L83
```
links[ i ] = new URL( url,  window._googlesitekitLegacyData.admin.siteURL).href.replace( window._googlesitekitLegacyData.admin.siteURL,'')
```

See how the scenario is avoided with the new widget based approach here:
https://github.com/google/site-kit-wp/blob/31d77ec6a4370be08ec67177156479535237f8fa/assets/js/components/DetailsPermaLinks.js#L39

**Comment by felixarntz** [comment]
@ivankruchkoff I wonder if it would be simpler to do it the other way around:
* Rather than stripping something from the path that Analytics API responses return, conditionally based on whether the base site URL includes a path, it might be easier to only prefix these URLs with the base site URL _without_ any path segment.
* The Analytics paths returned are always the full paths, relative to the domain/hostname. Instead of modifying every path returned, the approach I'm suggesting would only strip the path segment from the site URL once and then concatenate the result with the full path from the Analytics API response.

This way we might also get to a slightly smaller estimate I'd imagine. What do you think?

**Comment by ivankruchkoff** [comment]
@felixarntz 

The proposal doesn't conditionally exclude things and its implementation is mostly in the IB:

```js
function getURLPathExcludingSiteURL(path) {
  const siteURL = useSelect( ( select ) => select( CORE_SITE ).getReferenceSiteURL();
  return new URL(path, siteURL).href.replace(siteURL,'');
} 
```

Take the full path from analytics, combine it with the the site url, and then remove the site url. This lets us make minimal change in the legacy code without having to touch `getDataTableFromData`. We could if we wanted to not even remove the site url, and just do the following:

```js
function getFullURL(path) {
  const siteURL = useSelect( ( select ) => select( CORE_SITE ).getReferenceSiteURL();
  return new URL(path, siteURL);
} 
```

as the function doesn't change absolute links:
https://github.com/google/site-kit-wp/blob/31d77ec6a4370be08ec67177156479535237f8fa/assets/js/components/data-table.js#L53

The reason for the estimate is that I foresee quite a bit of QA time to make sure it all works correctly, and to get the repro scenarios in place. I estimate engineering portion is only a 3 but we don't separate QA from Eng. Happy to bring the overall ticket down to a 7 or 11 if we need to.

**Comment by felixarntz** [comment]
@ivankruchkoff I think I see where our misunderstanding is coming from now; your approach for the IB goes further than what is specified in the ACs:
* The ACs only refer to changing the behavior for when the site URL is combined with a path. This is the case for example in the full site URLs that the data tables may link to.
* Your IB focuses on stripping the path prefix from the path if it is already included in the site URL. In a way this could also make sense, but since we are actually displaying full paths, I don't think the currently displayed values in the data table are unexpected.

In other words, rather than creating an alternative to `getURLPath`, we should be focusing on fixing how to get the correct _full_ URL based on a path, which is where the problem specified in the ACs is. For example, instead of `siteURL + path` or `new URL( path, siteURL ).href`, we should do something like `getFullURL( siteURL, path )` where then `getFullURL` would take care of stripping the common path and return the full URL. So this issue's focus is only about changing the full URLs which are incorrect, not change individual paths and how they are displayed.

**Comment by ivankruchkoff** [comment]
@felixarntz  we are talking about the same thing, just different ways of implementing it. The end result is changing href urls to not contain the path twice, and in no way updating the displayed path as that is already correct.

I'll update to `getFullURL( siteURL, path )` which will do absolute links in the data-table.js and not prepend the link with the siteurl.

**Comment by felixarntz** [comment]
@ivankruchkoff I think two things are still off with the implementation here: Most code you're referencing in the IB deals with URL paths only, and those are all okay as is. The problem outlined in the ACs only occurs when concatenating the site URL with a URL path.

> Update the url to use the new function in the following places:
https://github.com/google/site-kit-wp/blob/31d77ec6a4370be08ec67177156479535237f8fa/assets/js/modules/analytics/components/dashboard/LegacyAnalyticsDashboardWidgetPopularPagesTable.js#L83
https://github.com/google/site-kit-wp/blob/31d77ec6a4370be08ec67177156479535237f8fa/assets/js/modules/analytics/components/dashboard/LegacyAdSenseDashboardWidgetTopPagesTableSmall.js#L87

You're proposing to change the components that use `DataTable`, however these two components only deal with the paths, so they are not where the problem is coming from. The problem lies in `DataTable` itself where it just concatenates the site URL with the path (see https://github.com/google/site-kit-wp/blob/31d77ec6a4370be08ec67177156479535237f8fa/assets/js/components/data-table.js#L53). So that line needs to instead use `getFullURL`.

> Check that the new widgets work correctly:
https://github.com/google/site-kit-wp/blob/31d77ec6a4370be08ec67177156479535237f8fa/assets/js/modules/adsense/components/dashboard/DashboardTopEarningPagesWidget.js#L106
https://github.com/google/site-kit-wp/blob/31d77ec6a4370be08ec67177156479535237f8fa/assets/js/modules/analytics/components/dashboard/DashboardPopularPagesWidget.js#L125
https://github.com/google/site-kit-wp/blob/31d77ec6a4370be08ec67177156479535237f8fa/assets/js/components/wp-dashboard/WPDashboardPopularPages.js#L104

You're proposing to change the components that use `DetailsPermaLinks`, but these components as well only deal with paths. This part should already work correctly (as `DetailsPermaLinks` already uses the same code as `getFullURL` would), so I think all you need to do here is update `DetailsPermaLinks` to use the new function (instead of manually containing the same code).

**Comment by felixarntz** [comment]
IB ✅

**Comment by wpdarren** [comment]
@cole10up @aaemnnosttv do you have a site with a subdirectory install that has analytics data? I've tried changing the directory of a Jurassic ninja site and it keeps erroring out. I have a live domain but it does not have enough data to test this.

**Comment by aaemnnosttv** [comment]
@jamesozzie @abdullah1908 looks like you were able to reproduce this – can you connect with @wpdarren to help him test it?

**Comment by abdullah1908** [comment]
@aaemnnosttv I'll follow-up with @wpdarren on this.

**Comment by wpdarren** [comment]
Site set up for this. Just waiting for Analytics data to update.

**Comment by wpdarren** [comment]
QA Update:  Pass ✅ 
 
**Verified:**

- The actual data appears on the Detailed Page Stats and the error message does not appear. 

![image](https://user-images.githubusercontent.com/73545194/117155107-667b3080-adb4-11eb-9042-082c52aa3360.png)

- The url in the analytics doesn't have the folder twice.

![image](https://user-images.githubusercontent.com/73545194/117156116-45670f80-adb5-11eb-9c5b-53e8c0c68bc6.png)

![image](https://user-images.githubusercontent.com/73545194/117156189-57e14900-adb5-11eb-889d-72e00024bcb5.png)
